File: internal/binmanager/binmanager.go
package binmanager

import (
	"archive/zip"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"

	"github.com/iamnotpirates/idlixdownloader/internal/config"
)

type BinPaths struct {
	FFmpeg    string
	NM3U8DLRE string
}

type BinManager struct {
	binDir string
}

func New() *BinManager {
	return &BinManager{
		binDir: config.GetBinDir(),
	}
}

func (bm *BinManager) GetNM3U8DLREPath() string {
	exeName := "N_m3u8DL-RE"
	if runtime.GOOS == "windows" {
		exeName += ".exe"
	}

	// 1. Check local AppData bin dir
	appDataPath := filepath.Join(bm.binDir, exeName)
	if fileExists(appDataPath) {
		return appDataPath
	}

	// 2. Check D:\Tools\idlixdownloader\bin
	toolsPath := filepath.Join("D:\\Tools\\idlixdownloader\\bin", exeName)
	if fileExists(toolsPath) {
		return toolsPath
	}

	// 3. Check system PATH
	if p, err := exec.LookPath(exeName); err == nil {
		return p
	}

	return ""
}

func (bm *BinManager) GetFFmpegPath() string {
	exeName := "ffmpeg"
	if runtime.GOOS == "windows" {
		exeName += ".exe"
	}

	// 1. Check local AppData bin dir
	appDataPath := filepath.Join(bm.binDir, exeName)
	if fileExists(appDataPath) {
		return appDataPath
	}

	// 2. Check D:\Tools\idlixdownloader\bin
	toolsPath := filepath.Join("D:\\Tools\\idlixdownloader\\bin", exeName)
	if fileExists(toolsPath) {
		return toolsPath
	}

	// 3. Check system PATH
	if p, err := exec.LookPath(exeName); err == nil {
		return p
	}

	return ""
}

func (bm *BinManager) EnsureBinaries() (*BinPaths, error) {
	_ = os.MkdirAll(bm.binDir, 0755)

	nPath := bm.GetNM3U8DLREPath()
	if nPath == "" {
		fmt.Println("[BinManager] N_m3u8DL-RE not found. Auto-downloading...")
		var err error
		nPath, err = bm.downloadNM3U8DLRE()
		if err != nil {
			return nil, fmt.Errorf("failed to download N_m3u8DL-RE: %w", err)
		}
	} else if filepath.Dir(nPath) != bm.binDir {
		// Copy to AppData bin dir if loaded from external location
		target := filepath.Join(bm.binDir, filepath.Base(nPath))
		if !fileExists(target) {
			_ = copyBinary(nPath, target)
			nPath = target
		}
	}

	fPath := bm.GetFFmpegPath()
	if fPath == "" {
		fmt.Println("[BinManager] ffmpeg not found. Auto-downloading...")
		var err error
		fPath, err = bm.downloadFFmpeg()
		if err != nil {
			return nil, fmt.Errorf("failed to download ffmpeg: %w", err)
		}
	} else if filepath.Dir(fPath) != bm.binDir {
		// Copy to AppData bin dir if loaded from external location
		target := filepath.Join(bm.binDir, filepath.Base(fPath))
		if !fileExists(target) {
			_ = copyBinary(fPath, target)
			fPath = target
		}
	}

	return &BinPaths{
		FFmpeg:    fPath,
		NM3U8DLRE: nPath,
	}, nil
}

func copyBinary(src, dst string) error {
	in, err := os.Open(src)
	if err != nil {
		return err
	}
	defer in.Close()

	out, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer out.Close()

	_, err = io.Copy(out, in)
	return err
}

func (bm *BinManager) downloadNM3U8DLRE() (string, error) {
	client := &http.Client{Timeout: 120 * time.Second}
	req, _ := http.NewRequest("GET", "https://api.github.com/repos/nilaoda/N_m3u8DL-RE/releases/latest", nil)
	req.Header.Set("User-Agent", "IDLIXDownloader-Go/1.0")

	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	var release struct {
		Assets []struct {
			Name               string `json:"name"`
			BrowserDownloadURL string `json:"browser_download_url"`
		} `json:"assets"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&release); err != nil {
		return "", err
	}

	var downloadURL string
	for _, a := range release.Assets {
		if strings.Contains(strings.ToLower(a.Name), "win-x64") && strings.HasSuffix(a.Name, ".zip") {
			downloadURL = a.BrowserDownloadURL
			break
		}
	}

	if downloadURL == "" {
		return "", fmt.Errorf("compatible N_m3u8DL-RE release asset not found")
	}

	zipPath := filepath.Join(bm.binDir, "nm3u8dl.zip")
	if err := downloadFile(downloadURL, zipPath); err != nil {
		return "", err
	}
	defer os.Remove(zipPath)

	if err := extractZipBinary(zipPath, "N_m3u8DL-RE.exe", filepath.Join(bm.binDir, "N_m3u8DL-RE.exe")); err != nil {
		return "", err
	}

	return filepath.Join(bm.binDir, "N_m3u8DL-RE.exe"), nil
}

func (bm *BinManager) downloadFFmpeg() (string, error) {
	downloadURL := "https://github.com/GyanD/codexffmpeg/releases/download/7.1/ffmpeg-7.1-essentials_build.zip"
	zipPath := filepath.Join(bm.binDir, "ffmpeg.zip")

	if err := downloadFile(downloadURL, zipPath); err != nil {
		return "", err
	}
	defer os.Remove(zipPath)

	if err := extractZipBinary(zipPath, "ffmpeg.exe", filepath.Join(bm.binDir, "ffmpeg.exe")); err != nil {
		return "", err
	}

	return filepath.Join(bm.binDir, "ffmpeg.exe"), nil
}

func fileExists(path string) bool {
	info, err := os.Stat(path)
	if err != nil {
		return false
	}
	return !info.IsDir()
}

func downloadFile(url, dest string) error {
	client := &http.Client{Timeout: 180 * time.Second}
	resp, err := client.Get(url)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("bad HTTP status: %s", resp.Status)
	}

	out, err := os.Create(dest)
	if err != nil {
		return err
	}
	defer out.Close()

	_, err = io.Copy(out, resp.Body)
	return err
}

func extractZipBinary(zipPath, targetName, outPath string) error {
	r, err := zip.OpenReader(zipPath)
	if err != nil {
		return err
	}
	defer r.Close()

	for _, f := range r.File {
		if filepath.Base(f.Name) == targetName {
			rc, err := f.Open()
			if err != nil {
				return err
			}
			defer rc.Close()

			outFile, err := os.Create(outPath)
			if err != nil {
				return err
			}
			defer outFile.Close()

			_, err = io.Copy(outFile, rc)
			return err
		}
	}
	return fmt.Errorf("file %s not found in zip archive", targetName)
}

File: backup/src/bin_manager.rs
use serde_json::Value;
use std::fs::{self, File};
use std::io::{self, Cursor};
use std::path::PathBuf;
use zip::ZipArchive;

#[derive(Debug, Clone)]
pub struct BinPaths {
    pub ffmpeg: PathBuf,
    pub n_m3u8dl_re: PathBuf,
}

pub struct BinManager {
    bin_dir: PathBuf,
}

impl BinManager {
    pub fn new() -> Self {
        let bin_dir = std::env::current_exe()
            .ok()
            .and_then(|p| p.parent().map(|p| p.to_path_buf()))
            .unwrap_or_else(|| PathBuf::from("."))
            .join("bin");

        Self { bin_dir }
    }

    /// Checks if a command binary is in system PATH
    fn find_in_path(name: &str) -> Option<PathBuf> {
        let exe_name = if cfg!(windows) && !name.ends_with(".exe") {
            format!("{name}.exe")
        } else {
            name.to_string()
        };

        if let Ok(path_var) = std::env::var("PATH") {
            for dir in std::env::split_paths(&path_var) {
                let full = dir.join(&exe_name);
                if full.is_file() {
                    return Some(full);
                }
            }
        }
        None
    }

    /// Checks if binary exists in local ./bin/ or PATH
    pub fn get_n_m3u8dl_re_path(&self) -> Option<PathBuf> {
        let local = self.bin_dir.join(if cfg!(windows) { "N_m3u8DL-RE.exe" } else { "N_m3u8DL-RE" });
        if local.is_file() {
            return Some(local);
        }
        Self::find_in_path("N_m3u8DL-RE")
    }

    /// Checks if ffmpeg exists in local ./bin/ or PATH
    pub fn get_ffmpeg_path(&self) -> Option<PathBuf> {
        let local = self.bin_dir.join(if cfg!(windows) { "ffmpeg.exe" } else { "ffmpeg" });
        if local.is_file() {
            return Some(local);
        }
        Self::find_in_path("ffmpeg")
    }

    /// Ensures both required binaries are available. Downloads them automatically if missing.
    pub async fn ensure_binaries(&self) -> Result<BinPaths, String> {
        fs::create_dir_all(&self.bin_dir)
            .map_err(|e| format!("Failed to create bin directory: {e}"))?;

        let n_path = match self.get_n_m3u8dl_re_path() {
            Some(p) => p,
            None => {
                println!("[BinManager] N_m3u8DL-RE not found. Auto-downloading...");
                self.download_n_m3u8dl_re().await?
            }
        };

        let ffmpeg_path = match self.get_ffmpeg_path() {
            Some(p) => p,
            None => {
                println!("[BinManager] ffmpeg not found. Auto-downloading...");
                self.download_ffmpeg().await?
            }
        };

        Ok(BinPaths {
            ffmpeg: ffmpeg_path,
            n_m3u8dl_re: n_path,
        })
    }

    async fn download_n_m3u8dl_re(&self) -> Result<PathBuf, String> {
        let client = reqwest::Client::builder()
            .timeout(std::time::Duration::from_secs(120))
            .user_agent("IDLIXDownloader-Rust/1.0")
            .build()
            .map_err(|e| format!("HTTP error: {e}"))?;

        // 1. Fetch latest release URL from GitHub API
        let mut download_url = None;
        if let Ok(resp) = client
            .get("https://api.github.com/repos/nilaoda/N_m3u8DL-RE/releases/latest")
            .send()
            .await
        {
            if let Ok(rel) = resp.json::<Value>().await {
                if let Some(assets) = rel["assets"].as_array() {
                    for a in assets {
                        let name = a["name"].as_str().unwrap_or("");
                        let u = a["browser_download_url"].as_str().unwrap_or("");
                        if cfg!(windows) && name.contains("win-x64") && name.ends_with(".zip") {
                            download_url = Some(u.to_string());
                            break;
                        }
                    }
                }
            }
        }

        let url = download_url.unwrap_or_else(|| {
            "https://github.com/nilaoda/N_m3u8DL-RE/releases/download/v0.6.0-beta/N_m3u8DL-RE_v0.6.0-beta_win-x64_20260629.zip".to_string()
        });

        let resp = client
            .get(&url)
            .send()
            .await
            .map_err(|e| format!("Failed to download N_m3u8DL-RE: {e}"))?;

        if !resp.status().is_success() {
            return Err(format!("Download failed with status {}", resp.status()));
        }

        let bytes = resp.bytes().await.map_err(|e| format!("Read bytes error: {e}"))?;
        let target_exe = self.bin_dir.join(if cfg!(windows) { "N_m3u8DL-RE.exe" } else { "N_m3u8DL-RE" });

        // Extract zip
        let mut archive = ZipArchive::new(Cursor::new(bytes))
            .map_err(|e| format!("Invalid zip archive: {e}"))?;

        for i in 0..archive.len() {
            let mut file = archive.by_index(i).map_err(|e| format!("Zip entry error: {e}"))?;
            let name = file.name().to_string();
            if name.ends_with("N_m3u8DL-RE.exe") || name == "N_m3u8DL-RE" {
                let mut out = File::create(&target_exe).map_err(|e| format!("Failed to create exe: {e}"))?;
                io::copy(&mut file, &mut out).map_err(|e| format!("Copy failed: {e}"))?;
                break;
            }
        }

        if target_exe.is_file() {
            println!("[BinManager] N_m3u8DL-RE installed to {:?}", target_exe);
            Ok(target_exe)
        } else {
            Err("Failed to extract N_m3u8DL-RE binary from zip".to_string())
        }
    }

    async fn download_ffmpeg(&self) -> Result<PathBuf, String> {
        let target_exe = self.bin_dir.join(if cfg!(windows) { "ffmpeg.exe" } else { "ffmpeg" });

        let url = "https://github.com/yt-dlp/FFmpeg-Builds/releases/download/latest/ffmpeg-master-latest-win64-gpl.zip";

        let client = reqwest::Client::builder()
            .timeout(std::time::Duration::from_secs(180))
            .user_agent("IDLIXDownloader-Rust/1.0")
            .build()
            .map_err(|e| format!("HTTP error: {e}"))?;

        let resp = client
            .get(url)
            .send()
            .await
            .map_err(|e| format!("Failed to download ffmpeg: {e}"))?;

        if !resp.status().is_success() {
            return Err(format!("Download ffmpeg failed with status {}", resp.status()));
        }

        let bytes = resp.bytes().await.map_err(|e| format!("Read ffmpeg bytes error: {e}"))?;

        let mut archive = ZipArchive::new(Cursor::new(bytes))
            .map_err(|e| format!("Invalid ffmpeg zip: {e}"))?;

        for i in 0..archive.len() {
            let mut file = archive.by_index(i).map_err(|e| format!("Zip entry error: {e}"))?;
            let name = file.name().to_string();
            if name.ends_with("bin/ffmpeg.exe") || name.ends_with("ffmpeg.exe") {
                let mut out = File::create(&target_exe).map_err(|e| format!("Failed to create ffmpeg.exe: {e}"))?;
                io::copy(&mut file, &mut out).map_err(|e| format!("Copy ffmpeg failed: {e}"))?;
                break;
            }
        }

        if target_exe.is_file() {
            println!("[BinManager] ffmpeg installed to {:?}", target_exe);
            Ok(target_exe)
        } else {
            Err("Failed to extract ffmpeg binary from zip".to_string())
        }
    }
}
